### ISSUE TITLE
nixos/nextcloud: do not default to sqlite

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#355606]
## Issue description

The nextcloud NixOS module currently defaults to sqlite as the database implementation for Nextcloud. This is problematic however because upstream Nextcloud does not appear to treat sqlite as a production-ready DB backend, thereby giving it little to no support.

https://docs.nextcloud.com/server/latest/admin_manual/installation/system_requirements.html only recommends SQLite for "testing and minimal-instances" and https://docs.nextcloud.com/server/latest/admin_manual/configuration_database/linux_database_configuration.html doesn't mention SQLite at all.

I've already encountered a breaking issue pertaining to the database with a relatively basic instance when trying to import my contacts and apparently this is ["expected behaviour"](https://help.nextcloud.com/t/import-partially-failed-general-error-5-database-is-locked/188032).

This issue is further exacerbated by the fact that [it hasn't been possible to convert databases for months](https://github.com/nextcloud/server/issues/45257) and there's no sign of that changing any time soon.

If such a basic task is enough to break nextcloud's SQLite support, I wouldn't consider it to be usable to a degree where it could reasonably be the default DB implementation for the NixOS module, especially if there's no way to reasonably switch to something else.

I can't really comment on which of mysql or postgres would be the better alternative but I'd gravitate toward postgres personally.

cc @Ma27 @SuperSandro2000 @onny

**Comment by SuperSandro2000** [comment]
Do we need to change more than https://github.com/NixOS/nixpkgs/blob/76612b17c0ce71689921ca12d9ffdc9c23ce40b2/nixos/modules/services/web-apps/nextcloud.nix#L420 to pgsql? Probably enable postgresql itself?

**Comment by Atemu** [comment]
There's already an option to create the DBs locally.

The reason I created this issue is to spread awareness and to discuss which DB to recommend instead. I'm not 100% certain it should be postgres because upstream's recommendations appear to favour mysql/mariadb.

We should also ensure that the default values offer a good user experience:

1. Document the fact that a DB must be provided in the enable option
2. Ensure the eval errors if you don't tell you the right things

**Comment by malikwirin** [comment]
> This issue is further exacerbated by the fact that [it hasn't been possible to convert databases for months](https://github.com/nextcloud/server/issues/45257) and there's no sign of that changing any time soon.

So this is a form of vendor lock-in?

**Comment by Ma27** [comment]
Agreed on the sqlite-default situation. We may want to revisit the phpfpm settings as well, they're IIRC pretty conservative here (I increased the worker settings a long while ago on my three-user-instance e.g.).

For me it wasn't such a big deal, but I wasn't aware of the situation about switching DBs makes it clear that we should provide a more reasonable default, thanks for bringing this to my attention @atemu 

>  to pgsql?

I'm wondering if this is a case for `system.stateVersion` here: this is a rather intrusive change and we don't have a reasonable way to warn.

> Probably enable postgresql itself?

Only when using a local socket path I'd say. I don't think it's nice having to turn off a postgresql on each node with a service talking to a postgresql instance.

> I'm not 100% certain it should be postgres because upstream's recommendations appear to favour mysql/mariadb.

The reason why the PHP ecosystem generally prefers MySQL/MariaDB/Percona is that establishing connections is far cheaper than on postgresql and in case of PHP this matters where each request bootstraps the framework, the DB connection etc. (there are ways around this with e.g. pgbadger).

OTOH I recall incompatibilities with certain MariaDB versions that didn't get fixed for months whereas I've seen Nextcloud installations with PostgreSQL being fine over years.

I'm not sure yet which I'd prefer, but I agree that we should change the default.

**Comment by Atemu** [comment]
> I'm wondering if this is a case for `system.stateVersion` here: this is a rather intrusive change

Yeah existing installations should remain on sqlite and stateVersion of current release + 1 should default to the alternative.

> we don't have a reasonable way to warn.

We could show a warning in the default of `dbtype` if we're in an affected stateVersion. That way users who are on the default would see it and those who explicitly set `dbtype = "sqlite"` will not see it.

What I have more issue with is what we'd even tell the user in the warning. "Hey ur db sux but you can't fix it, kthxbye" wouldn't be very helpful ^^'

> The reason why the PHP ecosystem generally prefers MySQL/MariaDB/Percona is that establishing connections is far cheaper than on postgresql and in case of PHP this matters where each request bootstraps the framework, the DB connection etc. (there are ways around this with e.g. pgbadger).
> 
> OTOH I recall incompatibilities with certain MariaDB versions that didn't get fixed for months whereas I've seen Nextcloud installations with PostgreSQL being fine over years.

Always good to remind myself that there's a reason I don't touch PHP.

This sounds to me like postgres would be the sane low-maintenance default and admins of larger instances could opt to use mariadb should they need it for performance reasons. This maps to how redis/memcached are optional. Also keeping in mind that premature optimisation is the root of all evil, I don't see any great need to have mariadb be the default.

**Comment by SuperSandro2000** [comment]
NixOS generally prefers postgres over mariabd so I would just go with that and in combination to that Ma and me run it with postgres.
